src/common/ui/codemirror-js-mixed/javascript-mixed.js

```diff
@@ -0,0 +1,756 @@
+      const quoteInUse = state.quoteCharSurroundJsExpr;
+      // first ensure, we let the css tokenizer continue the next time
+      state.tokenizePostJsExpr = null;
+      state.quoteCharSurroundJsExpr = null;
+
+      if (!quoteInUse) {
+        return null;
+      }
+
+      // Now handle quoted string cases such as content: "suffix${someExpr()}prefix";
+      // to return prefix" as a string token in the above case
+      // regex: non-greedy match up to the immediate next quote char, to avoid over match
+      const matched = stream.match(new RegExp(`.*?${quoteInUse}`), true);
+      dbg('  css mode post js expr string token - ', stream.current());
+      // in the unexpected case (likely bugs) that we cannot find end quote, do nothing more
+      // and let parent mode tokenizer to do its work
+      return matched ? 'string' : null;
+    };
+
+    function indexOfJsExprStartInString(text) {
+      // whether encounter js expression ${, but excluding cases that $ is escaped, i.e., \$
+      const jsExprMatch = text.match(/[^\\][$][{]|^[$][{]/);
+      if (!jsExprMatch) {
+        return -1;
+      }
+      // case there is an js expression
+      return jsExprMatch[0].charAt(0) === '$' ? jsExprMatch.index : jsExprMatch.index + 1;
+    }
+
+    function tokenJsExpressionInStringTemplate(stream, state) {
+      const style = jsMode.token(stream, state.jsState);
+      dbg('  local mode js expr token - ', stream.current(), `[${style}]`, state.jsState.lastType);
+      // track ${ , } to determine when the expression is complete.
+      const text = stream.current();
+      if (style === 'string-2' && indexOfJsExprStartInString(text) >= 0) { // case nested ${
+        state.jsExprDepthInStringTemplate += 1;
+        dbg('    jsExprDepthInStringTemplate inc:', state.jsExprDepthInStringTemplate);
+      } else if (style === 'string-2' && state.jsState.lastType === '}') { // case expression-ending }
+        // Note: must check BOTH style and lastType.
+        // If there are blank spaces after },
+        // when tokenizing the blank spaces, the style is null but the lastType remains to be }
+        // (the one with meaningful token)
+
+        // once it reaches back to 0, the logic would let the parent local mode handle the next token
+        state.jsExprDepthInStringTemplate -= 1;
+        dbg('    jsExprDepthInStringTemplate dec:', state.jsExprDepthInStringTemplate);
+        if (state.jsExprDepthInStringTemplate <= 0) {
+          state.inJsExprInStringTemplate = false;
+          if (state.localMode.tokenizePostJsExpr) {
+            // unless the mode also explicitly specify a tokenizer.
+            state.tokenizePostJsExpr = state.localMode.tokenizePostJsExpr;
+          }
+        }
+      }
+      return style;
+    }
+
+    function indexOfBacktickInString(text) {
+      // whether encounter backtick ` (not those escaped)
+      const jsExprMatch = text.match(/[^\\][`]|^[`]/);
```

Comment:
The first part ```[^\\][`]``` can be anywhere, not just at the beginning, is this intended?

Reply:
It's intentaional: the logic is to match backtick anywhere, but not those that are escaped, i.e., <code>\\`</code>.
 
The second part, <code>^[`]</code>, is to handle cases that the backtick happen to be the first character.

Examples for the match logic
```javascript
  someHTML=`<p>text</p>more text post last tag`;
```
After handling `</p>`, the html tokenizer would next tokenize <code>more text post last tag`;</code>, not knowing it is running in a JS string template and thus the backtick is meant to end the string template (and html). The logic here is to handle such cases.